### PR TITLE
Wiki specs

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: doc/specs/wiki.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
+# Wiki Specs
+
+## Metadata
+
+Author: guitarman0831
```
Comment (ghost) [inline doc/specs/wiki.md]: Not including author fields

File: doc/specs/wiki.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+### Changelog
```
Comment (ghost) [inline doc/specs/wiki.md]: Wasn't including the change log either

File: doc/specs/wiki.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+4. Data model
+
+## Motivation
+
+>The site contains a variety of users with a vast wealth of knowledge. The wiki is a place where users can put their knowledge to work, and write articles for users to use. These may be supplementary to other documents on the site, or standalone documents such as guides.
+
+## Business rules
+
+> * Articles must be able to be read/write locked based on user class.
+> * Staff members must have full access to articles.
```
Comment (ghost) [inline doc/specs/wiki.md]: Could be rephrased as users with the wiki management permissions have full access

File: doc/specs/wiki.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+> * Article creation must be associated with a permission.
```
Comment (ghost) [inline doc/specs/wiki.md:18]: Associated with what permission? user class? If so that's been said

File: doc/specs/wiki.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+## Use cases
+
+These use cases describe various ways the authentication system will interact.
```
Comment (ghost) [inline doc/specs/wiki.md]: Does this belong here?

File: doc/specs/wiki.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+**Use cases:**
+
+### 1.0 Creating an article
+
+**Primary Actor:** A user
+
+**Brief:** Users must be able to create wiki articles in order to share knowledge with other users.
+
+**Preconditions:** The user has the proper permission to create a wiki article.
```
Comment (ghost) [inline doc/specs/wiki.md]: must not have the wiki creation permissions disabled or they have the wiki management permissions

File: doc/specs/wiki.md [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+**Basic Flow:**
+
+1. The user clicks the link to create an article.
+2. The user fills out the title and body of the new article.
+3. The user chooses the user class read and edit restrictions.
+4. The user clicks submit, and is redirected to the new article.
+
+**Functional Requirements**
+
+* BBCode should be stored in the database and rendered at page load
```
Comment (ghost) [inline doc/specs/wiki.md]: Non functional. Correct me if Im wrong

File: doc/specs/wiki.md [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+### 2.0 Viewing an article
+
+**Primary Actor:** A user
+
+**Brief:** Wiki articles must be able to be viewed.
+
+**Preconditions:** The user has the appropriate user class to read the article.
```
Comment (ghost) [inline doc/specs/wiki.md]: See other permission comments

File: doc/specs/wiki.md [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+**Basic Flow:**
+
+1. The user navigates to the URL of the article.
+2. The article is rendered.
+
+**Functional Requirements:**
+
+* BBCode must be rendered when the article is loaded from the database.
+* The table of contents must be generated based on the pre-defined BBCode header tags.
```
Comment (ghost) [inline doc/specs/wiki.md]: We don't enforce a table of contents.

File: doc/specs/wiki.md [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+* The article is rendered.
+
+**Basic Flow:**
+
+1. The user navigates to the URL of the article.
+2. The article is rendered.
+
+**Functional Requirements:**
+
+* BBCode must be rendered when the article is loaded from the database.
```
Comment (ghost) [inline doc/specs/wiki.md]: Implied

File: doc/specs/wiki.md [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+* The table of contents must be generated based on the pre-defined BBCode header tags.
+
+### 3.0 Editing an article
+
+**Primary Actor:** A user
+
+**Brief:** Editing articles allows mistakes and misinformation to be corrected, and current information to be expanded upon.
+
+**Preconditions:** The user has the appropriate user class to edit the article.
```
Comment (ghost) [inline doc/specs/wiki.md]: See other permission comments

File: doc/specs/wiki.md [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+**Trigger:** The user clicks the edit link on the article's page.
+
+**Postcondition:**
+
+* The edits to the article are saved.
+
+**Basic Flow:**
+
+1. The user clicks the edit link on the article's page.
+2. The user makes the appropriate changes to the title, body, and user class read/edit restrictions.
```
Comment (ghost) [inline doc/specs/wiki.md]: A typical user doesnt have access to user class read/write permissions

File: doc/specs/wiki.md [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+**Basic Flow:**
+
+1. The user is making changes to an article, or writing a new one.
+2. The preview button is clicked.
+3. The article is rendered in-place, and is presented to the user.
+
+**Functional Requirements:**
+
+* BBCode must be rendered when the article is previewed.
```
Comment (ghost) [inline doc/specs/wiki.md]: IMplied

File: doc/specs/wiki.md [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+* The alias is added to the article.
+
+**Basic Flow:**
+
+1. The user enters the new alias into the textbox.
+2. The user hits submit, and the alias is added to the article.
+
+**Functional Requirements:**
+
+* Duplicate and null aliases must be able to be detected, and discarded.
```
Comment (ghost) [inline doc/specs/wiki.md]: You need to elaborate on this

Comment (ghost) [inline doc/specs/wiki.md]: How do aliases fit into wiki search? thats another use case

File: doc/specs/wiki.md [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+**Basic Flow:**
+
+1. The user clicks the X button next to an alias.
+2. The alias is removed.
+
+**Functional Requirements:**
+
+* The alias must be removed from the database.
+
+### 4.2 Viewing a user that added an alias
```
Comment (ghost) [inline doc/specs/wiki.md]: Change this use case into "a user can view the wiki's aliases" 

Also how do aliases fit in with wiki permissions?

File: doc/specs/wiki.md [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+
+**Wiki Articles:**
+> * The title of the article.
+> * The body of the article, containing the BBCode tags.
+> * The current revision number.
+> * The ID of the user that created the article.
+> * The timestamp of the article creation time.
+> * The timestamp of the last edit time.
+
+**Wiki Revisions:**
```
Comment (ghost) [inline doc/specs/wiki.md]: Need to explore how to handle revisions

File: doc/specs/wiki.md [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+2. The user is redirected to the user that added the alias.
+
+## Data model
+
+What data should be tracked.
+
+**Wiki Articles:**
+> * The title of the article.
+> * The body of the article, containing the BBCode tags.
+> * The current revision number.
```
Comment (ghost) [inline doc/specs/wiki.md]: See revision comment

File: doc/specs/wiki.md [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+1. The user clicks the U button next to an alias.
+2. The user is redirected to the user that added the alias.
+
+## Data model
+
+What data should be tracked.
+
+**Wiki Articles:**
+> * The title of the article.
+> * The body of the article, containing the BBCode tags.
```
Comment (ghost) [inline doc/specs/wiki.md]: bbcode is implied

File: doc/specs/wiki.md (hunk 16 above)
Comment (ghost) [inline doc/specs/wiki.md]: Also this would be stored in the wiki document.

File: doc/specs/wiki.md [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+
+**Wiki Revisions:**
+> * The ID of the corresponding wiki article.
+> * The revision number.
+> * The revision title.
+> * The revision body.
+> * The timestamp of the revision date.
+> * The ID of the user that authored the revision.
+
+**Wiki Aliases:**
```
Comment (ghost) [inline doc/specs/wiki.md]: This would also be stored in the wiki document

File: doc/specs/wiki.md (hunk 1 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 2 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 3 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 5 above)
Comment (ghost) [inline doc/specs/wiki.md]: Nope. Resolved.

File: doc/specs/wiki.md (hunk 6 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 7 above)
Comment (ghost) [inline doc/specs/wiki.md]: I believe it's non functional. Resolved.

File: doc/specs/wiki.md (hunk 8 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 9 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 10 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 11 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 12 above)
Comment (ghost) [inline doc/specs/wiki.md]: I wasn't aware of this. Resolved.

File: doc/specs/wiki.md (hunk 13 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 15 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 16 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 17 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 18 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 19 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 14 above)
Comment (ghost) [inline doc/specs/wiki.md]: Reworded a little bit and added a search with aliases use case. Resolved.

File: doc/specs/wiki.md (hunk 4 above)
Comment (ghost) [inline doc/specs/wiki.md:18]: Some kind of wiki creation permission, similar to how invites or posting in the forums can be disabled.

I don't see any necessary change. Resolved.

File: doc/specs/wiki.md [hunk 20]
```diff
@@ -0,0 +1,411 @@
...
+
+**Functional Requirements:**
+
+* The article's revision history must be updated with the changes.
+
+**Non-Functional Requirements:**
+
+* The article editing template must be the same as the article creation template, except with all of the fields filled in.
+
+### 3.1 Previewing articles
```
Comment (ghost) [inline doc/specs/wiki.md]: Implied

File: doc/specs/wiki.md [hunk 21]
```diff
@@ -0,0 +1,411 @@
...
+
+**Basic Flow:**
+
+1. The user enters the new alias into the textbox.
+2. The user hits submit, and the alias is added to the article.
+
+**Functional Requirements:**
+
+* When the user hits submit, the system must ensure that the alias isn't a duplicate, and that it is not empty.
+* New aliases are committed to the database.
```
Comment (ghost) [inline doc/specs/wiki.md]: 1st one is a business rule, 2nd one is a post condition

File: doc/specs/wiki.md (hunk 20 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 21 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md [hunk 22]
```diff
@@ -0,0 +1,391 @@
...
+
+**Basic Flow:**
+
+1. The user clicks the link to create an article.
+2. The template is generated, containing a title and section textarea.
+3. The user fills out the fields and user class read/edit restrictions, and hits submit.
+
+**Non-Functional Requirements**
+
+* BBCode must be stored in the database and rendered at page load
```
Comment (ghost) [inline doc/specs/wiki.md]: All the bbcode loading/rendering is implied for the majority of the text, no need to include this.

File: doc/specs/wiki.md [hunk 23]
```diff
@@ -0,0 +1,391 @@
...
+* Each section must have a [+] button to add another section.
```
Comment (ghost) [inline doc/specs/wiki.md]: Omit the whole thing about [+] button, we don't know how it's going to look. Just say what is needed.

File: doc/specs/wiki.md [hunk 24]
```diff
@@ -0,0 +1,391 @@
...
+* Each section must have a [++] button to add a sub-section.
+
+### 2.0 Viewing an article
+
+**Primary Actor:** A user
+
+**Brief:** Wiki articles must be able to be viewed.
+
+**Preconditions:** The user has the appropriate user class to read the article, or has the wiki management permission enabled.
```
Comment (ghost) [inline doc/specs/wiki.md]: What if the user with the wiki management permission doesn't qualify the class requirements to view an article? Like a DT viewing an article set for Moderators?  Perhaps another permission is needed, or you could refine this condition. Consult the current wiki and see how it works.

File: doc/specs/wiki.md [hunk 25]
```diff
@@ -0,0 +1,391 @@
...
+
+**Basic Flow:**
+
+1. The user clicks the edit link on the article's page.
+2. The user clicks submit.
+3. The changes are saved to the database, and the user is redirected to the updated article.
+
+**Functional Requirements:**
+
+* The article's revision history must be updated with the changes.
```
Comment (ghost) [inline doc/specs/wiki.md:92]: This isn't a functional requirement, correct me I'm wrong

File: doc/specs/wiki.md (hunk 24 above)
Comment (ghost) [inline doc/specs/wiki.md]: Wiki permissions may need another layer. Really think about how this works. 

Also mention somewhere what needs to happen when a class is deleted which has wiki articles associated with it? We don't have FKs so gotta be extra careful.

File: doc/specs/wiki.md [hunk 26]
```diff
@@ -0,0 +1,391 @@
...
+
+**Brief:** Wiki articles must be able to be viewed.
+
+**Preconditions:** The user has the appropriate user class to read the article, or has the wiki management permission enabled.
+
+**Trigger:** The user navigates to the article's URL.
+
+**Postcondition:**
+
+* The article is rendered.
```
Comment (ghost) [inline doc/specs/wiki.md]: Open ended question, how much can we imply in these specs

File: doc/specs/wiki.md [hunk 27]
```diff
@@ -0,0 +1,391 @@
...
+2. The user clicks submit.
+3. The changes are saved to the database, and the user is redirected to the updated article.
+
+**Functional Requirements:**
+
+* The article's revision history must be updated with the changes.
+
+**Non-Functional Requirements:**
+
+* The article editing template must be the same as the article creation template, except with all of the fields filled in.
```
Comment (ghost) [inline doc/specs/wiki.md:93]: This is a functional requirement except it's something a user can do. A user can edit the title and body of the article

File: doc/specs/wiki.md [hunk 28]
```diff
@@ -0,0 +1,391 @@
...
+**Primary Actor:** A user
+
+**Brief:** Alises allow wiki articles to be easily searched. Managing aliases ensures they are accurate and thorough.
+
+**Preconditions:**
+
+* The user is currently viewing the article.
+* The user has the appropriate user class to edit the article.
+
+**Trigger:** The user clicks the submit button next to the alias textbox.
```
Comment (ghost) [inline doc/specs/wiki.md]: Unless necessary don't describe how it clicks, looks, etc. What if we don't have a button and just press enter? Keep this in mind throughout.

File: doc/specs/wiki.md [hunk 29]
```diff
@@ -0,0 +1,391 @@
...
+
+**Postcondition:**
+
+* The alias is added to the article and committed to the database.
```
Comment (ghost) [inline doc/specs/wiki.md]: What about revisions

File: doc/specs/wiki.md [hunk 30]
```diff
@@ -0,0 +1,391 @@
...
+
+**Basic Flow:**
+
+1. The user enters the new alias into the textbox.
+2. The user hits submit, and the alias is added to the article.
+
+**Applicable business rules:**
+
+### When the user hits submit, the system must ensure that the alias isn't a duplicate, and that it is not empty.
```
Comment (ghost) [inline doc/specs/wiki.md]: Move this to business rules and mention in the requiremetns

File: doc/specs/wiki.md [hunk 31]
```diff
@@ -0,0 +1,391 @@
...
+**Primary Actor:** A user
+
+**Brief:** Aliases must be able to ensure relevancy and accuracy.
+
+**Preconditions:**
+
+* The user is currently viewing the article.
+* The user has the appropriate user class to edit the article.
+
+**Trigger:** The user clicks the X button next to an alias.
```
Comment (ghost) [inline doc/specs/wiki.md]: You're killing independent George!

File: doc/specs/wiki.md (hunk 22 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 23 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 24 above)
Comment (ghost) [inline doc/specs/wiki.md]: @ajax-wcd I currently have 3 permissions: wiki creation, wiki management, and wiki admin. admins have full access to any wiki articles, management allows editing/deleting articles that the user has the proper class to view/edit, and creation allows users to create articles.

Edit: Make that 4 permissions, the 4th being commenting on wiki articles.

Resolved.

File: doc/specs/wiki.md (hunk 25 above)
Comment (ghost) [inline doc/specs/wiki.md:92]: @ajax-wcd I believe this is functional, seeing that it relates directly to the database.

File: doc/specs/wiki.md (hunk 26 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 27 above)
Comment (ghost) [inline doc/specs/wiki.md:93]: Resolved.

File: doc/specs/wiki.md (hunk 28 above)
Comment (ghost) [inline doc/specs/wiki.md]: Resolved.

File: doc/specs/wiki.md (hunk 30 above)
Comment (ghost) [inline doc/specs/wiki.md]: `Duplicate aliases, tags, and dedicated terms must be unique and not empty.`

It's already a business rule. Resolved.

File: doc/specs/wiki.md (hunk 31 above)
Comment (ghost) [inline doc/specs/wiki.md]: Lol, nice reference. Not sure what you mean by it but I removed all of the specifics related to UI elements.

Resolved.

File: doc/specs/wiki.md (hunk 29 above)
Comment (ghost) [inline doc/specs/wiki.md]: Managing aliases doesn't change revisions.

Resolved.

File: doc/specs/header.md [hunk 32]
```diff
@@ -0,0 +1,178 @@
...
+
+Author: sneakysneaky
+
+### Changelog
+
+**2015-07-13**
+
+* Initial version
+
+## Table of contects
```
Comment (pgoforth) [inline doc/specs/header.md]: "Contents" mispelled

Comment (Malvar) [inline doc/specs/header.md:13]: "Contents" mispelled

**Comment by ghost** [comment]
Issue #30

**Pull request closed without merging**